File: quantbit_pro_work_management/quantbit_pro_work_management/report/document_application_report/document_application_report.js
// Copyright (c) 2026, Quantbit Technologies Pvt. Ltd. and contributors
// For license information, please see license.txt

frappe.query_reports["Document Application Report"] = {
    filters: [
        {
            fieldname: "from_date",
            label: __("From Date"),
            fieldtype: "Date"
        },
        {
            fieldname: "to_date",
            label: __("To Date"),
            fieldtype: "Date"
        },
        {
            fieldname: "transaction_type",
            label: __("Application Type"),
            fieldtype: "Select",
            options: "\nNew Application\nRenewal\nExtension\nReplacement\nCancellation"
        },
        {
            fieldname: "applicant_full_name",
            label: __("Applicant Full Name"),
            fieldtype: "Data"
        },
        {
            fieldname: "status",
            label: __("Status"),
            fieldtype: "Select",
            options: "\nDraft\nSubmitted\nUnder Review\nWaiting Docs\nApproved\nRejected\nIssued\nActive\nRenewed\nExtended\nExpired\nCancelled"
        }
    ]
};

File: quantbit_pro_work_management/quantbit_pro_work_management/report/document_application_report/document_application_report.py
# Copyright (c) 2026, Quantbit Technologies Pvt. Ltd. and contributors
# For license information, please see license.txt

import frappe

def execute(filters=None):
    columns = get_columns()
    data = get_data(filters)
    return columns, data


def get_columns():
    return [
        {
            "label": "Application ID",
            "fieldname": "name",
            "fieldtype": "Link",
            "options": "Document Application",
            "width": 180,
        },
        {
            "label": "Applicant Full Name",
            "fieldname": "applicant_full_name",
            "fieldtype": "Data",
            "width": 180,
        },
        {
            "label": "Document Category",
            "fieldname": "document_category",
            "fieldtype": "Link",
            "options": "Document Category",
            "width": 160,
        },
        {
            "label": "Document Type",
            "fieldname": "document_type",
            "fieldtype": "Link",
            "options": "Document Type",
            "width": 160,
        },
        {
            "label": "Transaction Type",
            "fieldname": "transaction_type",
            "fieldtype": "Data",
            "width": 140,
        },
        {
            "label": "Posting Date",
            "fieldname": "posting_date",
            "fieldtype": "Date",
            "width": 120,
        },
        {
            "label": "Status",
            "fieldname": "status",
            "fieldtype": "Data",
            "width": 120,
        },
    ]


def get_data(filters):
    conditions = []
    values = {}

    if filters.get("from_date"):
        conditions.append("posting_date >= %(from_date)s")
        values["from_date"] = filters["from_date"]

    if filters.get("to_date"):
        conditions.append("posting_date <= %(to_date)s")
        values["to_date"] = filters["to_date"]

    if filters.get("transaction_type"):
        conditions.append("transaction_type = %(transaction_type)s")
        values["transaction_type"] = filters["transaction_type"]

    if filters.get("applicant_full_name"):
        conditions.append("applicant_full_name LIKE %(applicant_full_name)s")
        values["applicant_full_name"] = f"%{filters['applicant_full_name']}%"

    if filters.get("status"):
        conditions.append("status = %(status)s")
        values["status"] = filters["status"]

    condition_query = ""
    if conditions:
        condition_query = "WHERE " + " AND ".join(conditions)

    return frappe.db.sql(
        f"""
        SELECT
            name,
            applicant_full_name,
            document_category,
            document_type,
            transaction_type,
            posting_date,
            status
        FROM
            `tabDocument Application`
        {condition_query}
        ORDER BY posting_date DESC
        """,
        values,
        as_dict=True,
    )

